openedx/openedx-platform
# feat: incremental reindex_studio management command [FC-0062]

## Description

Adds an incremental mode to the reindex studio management command, and also other few utilities for managing the index: reset and init.

## Supporting information

* https://github.com/openedx/modular-learning/issues/235

## Testing instructions

First open the home page of any library. eg.: http://apps.local.openedx.io:2001/course-authoring/library/lib:test:2test

1. Using tutor, exec into the cms instance: `tutor dev exec cms bash`
2. Verify that the command requires the experimental flag: `./manage.py cms reindex_studio` should do nothing, and only reindex if you pass the `--experimental` flag
3. Test out the different flags for the command
    -  `./manage.py cms reindex_studio --experimental --init`: Should do nothing since the index already exist
    - `./manage.py cms reindex_studio --experimental --reset`: Should recreate an empty index. This means that searching in studio page should not return any results.
    - Run and interrupt `./manage.py cms reindex_studio --experimental --incremental` right after it finishes indexing collections. Run it again and assert it continues from where it was interrupted. Try searching content now.
    - Run  `./manage.py cms reindex_studio --experimental` and verify the index is recreated by searching blocks after it finishes. And also assert that there are no records of a current incremental update lingering: `./manage.py cms shell -c 'from openedx.core.djangoapps.content.search.models import IncrementalIndexCompleted; print(IncrementalIndexCompleted.objects.all())'`

## Deadline

Asap

## Other information

Private-Ref: https://tasks.opencraft.com/browse/FAL-3902
